feat: support docker image history scanning

jamestelfer opened the pull request:
### Description

> [!NOTE]
> This PR has been created to help illustrate a possible fix for #2881. I'd be grateful for any feedback, and if this is useful, I'll follow the guidance to ready it for merge.

Enables the `docker` scanner to send image history records to the scanner. These records are generally the commands that were run to build each image layer. The entries look similar to:

```json
    {
      "created": "2024-05-24T05:26:37.56186972Z",
      "created_by": "RUN |8 PNPM_VERSION=v8.15.5 GITHUB_REGISTRY_TOKEN=ghp_ohmygodnessaleakedvalue /bin/sh -c unwise-command '${GITHUB_REGISTRY_TOKEN}' ",
      "comment": "buildkit.dockerfile.v0"
    },
```

In this example, the secret is leaked from a build argument, but it could equally be from a command line literal.

Fixes #2881 

### Questions

1. I'm currently using `image://config-file/history/index` as the filename for each history entry ([see diff](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-ebf4fb4c65aea8ffed27f07423b39278c65970e7ea9e6905065745c3068313caR210)). Is this sufficient for identification? It doesn't look particularly flash, but I'm not sure what would be better.
2. The `Layer` in the Docker metadata is [entered as the empty string](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-ebf4fb4c65aea8ffed27f07423b39278c65970e7ea9e6905065745c3068313caR213). If this isn't OK, I'd welcome suggestions on fixing it.
3. I [added a new metric](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-7c3859b918ec2dbd4e563b9b02e631efc12fda1373dde17de99a8a58ed2c452aR19-R25) `docker_history_entries_scanned` for this, assuming this is the right thing to do.

### Checklist

* [x] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

jamestelfer committed:
feat: support docker image history scanning

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=2882) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=2882) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trufflesecurity/trufflehog?pullRequest=2882) it.</sub>

ahrav commented on pkg/sources/docker/docker_test.go:
```diff
@@ -110,3 +133,12 @@ func TestBaseAndTagFromImage(t *testing.T) {
 	}
 }
+
+func isHistoryChunk(t *testing.T, chunk *sources.Chunk) bool {
```
❤️

ahrav commented on pkg/sources/docker/docker.go:
```diff
@@ -181,6 +198,35 @@ func (s *Source) processImage(ctx context.Context, image string) (imageInfo, err
+
+	ctx.Logger().V(2).Info("scanning image history entry", "index", entryIndex)
+
+	if err := common.CancellableWrite(ctx, chunksChan, chunk); err != nil {
```
Suggestion: Since we aren't including additional context with the returned error, we can skip the `err != nil` check and directly `return common.CancellableWrite(ctx, chunksChan, chunk)`.

jamestelfer replied:
Thanks, will do. I often get caught in between leaving this error handling idiom and just returning directly, it's good to get a push to pick one :D

ahrav reviewed (approved):
Thanks for the addition @jamestelfer this looks great to me.

ahrav commented:
> ### Description
> Note
> 
> This PR has been created to help illustrate a possible fix for #2881. I'd be grateful for any feedback, and if this is useful, I'll follow the guidance to ready it for merge.
> 
> Enables the `docker` scanner to send image history records to the scanner. These records are generally the commands that were run to build each image layer. The entries look similar to:
> 
> ```json
>     {
>       "created": "2024-05-24T05:26:37.56186972Z",
>       "created_by": "RUN |8 PNPM_VERSION=v8.15.5 GITHUB_REGISTRY_TOKEN=ghp_ohmygodnessaleakedvalue /bin/sh -c unwise-command '${GITHUB_REGISTRY_TOKEN}' ",
>       "comment": "buildkit.dockerfile.v0"
>     },
> ```
> 
> In this example, the secret is leaked from a build argument, but it could equally be from a command line literal.
> 
> Fixes #2881
> 
> ### Questions
> 1. I'm currently using `image://config-file/history/index` as the filename for each history entry ([see diff](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-ebf4fb4c65aea8ffed27f07423b39278c65970e7ea9e6905065745c3068313caR210)). Is this sufficient for identification? It doesn't look particularly flash, but I'm not sure what would be better.
> 2. The `Layer` in the Docker metadata is [entered as the empty string](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-ebf4fb4c65aea8ffed27f07423b39278c65970e7ea9e6905065745c3068313caR213). If this isn't OK, I'd welcome suggestions on fixing it.
> 3. I [added a new metric](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-7c3859b918ec2dbd4e563b9b02e631efc12fda1373dde17de99a8a58ed2c452aR19-R25) `docker_history_entries_scanned` for this, assuming this is the right thing to do.
> 
> ### Checklist
> * [x]  Tests passing (`make test-community`)?
> * [x]  Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

I am not too well versed in our Docker source, but from what I can tell this looks pretty good to me. That being said, @zricethezav @joeleonjr would you mind giving this a look as well. Thanks.

rgmz commented:
> The Layer in the Docker metadata is [entered as the empty string](https://github.com/trufflesecurity/trufflehog/pull/2882/files#diff-ebf4fb4c65aea8ffed27f07423b39278c65970e7ea9e6905065745c3068313caR213). If this isn't OK, I'd welcome suggestions on fixing it.

It's possible to associate history with the layer ID by index (skipping empty layers).

For instance, if you look at the `postgres:latest` you can see:
1. `#(nop) ADD file:5a` is layer `254fddc33a7883e66303d20feb68795952cb91cb0af7574e129f6bfd8efad544`
2.  `RUN /bin/sh -c set -eux; \tgroupadd -r postgres --gid=999;` is layer `d3b78d99bcc3eee03f029078417d7d1076170980d4073719fdb6d8c187e003df`. 

**History**
```json
{
"history": [
        {
            "created": "2024-05-09T18:58:11Z",
            "created_by": "/bin/sh -c #(nop) ADD file:5aaace706aa00ff97d243daa2c29f5de88f124e1b97c570634f16eef90783286 in / "
        },
        {
            "created": "2024-05-09T18:58:11Z",
            "created_by": "/bin/sh -c #(nop)  CMD [\"bash\"]",
            "empty_layer": true
        },
        {
            "created": "2024-05-09T18:58:11Z",
            "created_by": "RUN /bin/sh -c set -eux; \tgroupadd -r postgres --gid=999; \tuseradd -r -g postgres --uid=999 --home-dir=/var/lib/postgresql --shell=/bin/bash postgres; \tmkdir -p /var/lib/postgresql; \tchown -R postgres:postgres /var/lib/postgresql # buildkit",
            "comment": "buildkit.dockerfile.v0"
        },
        ...
}
```

**Layers**
```json
{
        "Layers": [
            "254fddc33a7883e66303d20feb68795952cb91cb0af7574e129f6bfd8efad544/layer.tar",
            "d3b78d99bcc3eee03f029078417d7d1076170980d4073719fdb6d8c187e003df/layer.tar",
            "cfb5c12ea5e8675b8fbc75bf74b539422cd99ae65e6f1b3603638a2e6152e384/layer.tar",
            "7bd588371cf03e6283e8729a4a8e28c3e02864d6865eddcdd1c9a7c8718996be/layer.tar",
            ...
       ]
}
```

This can be validated by either `docker pull postgres && docker save postgres > layers.tar` or using a GUI like [Dive](https://github.com/wagoodman/dive/).
![image](https://github.com/trufflesecurity/trufflehog/assets/32133502/eb307612-f6cd-42b2-8ff3-7ba4fc5cd39d)

jamestelfer commented:
The association with layers like this is a great point.

I can annotate an entry with its corresponding layer without issue, as long as it's ok that some records will not have an associated layer. 

Not all entries will have a layer, but any may have a secret.

Given this, some records will have an empty string for the layer, and I presume that the proposed naming scheme is OK for a start?

jamestelfer commented on pkg/sources/docker/docker.go:
```diff
@@ -108,7 +108,24 @@ func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk, _ .
+			for idx, historyEntry := range config.History {
+				if err := s.processHistoryEntry(ctx, idx, historyEntry, imgInfo, chunksChan); err != nil {
+					scanErrs.Add(err)
+					return nil
```
Great question! I didn't consider this as I was writing it.

Looking into the call chain, the error propagated is from the channel write, so the right thing to do is finish early I think.

jamestelfer committed:
refactor: collapse error handling into return

Style suggestion from review feedback.

jamestelfer committed:
fix: associate layers with history entries

Where possible, add the associated layer to the history entry record. This may help tracing any issues discovered.

This also changes the entry reference format to `image-metadata:history:%d:created-by` which _may_ be more self-explanatory.

jamestelfer commented:
Thanks for all your feedback @ahrav and @rgmz! I think I've addressed it, looking forward to a re-review.

Note that I changed the entry reference format to `image-metadata:history:%d:created-by` which _may_ be more self-explanatory than `image://config-file/history/%d`.

ahrav reviewed (approved):
LGTM! Thanks again for adding this change.

jamestelfer commented on pkg/sources/docker/docker.go:
```diff
@@ -181,6 +206,86 @@ func (s *Source) processImage(ctx context.Context, image string) (imageInfo, err
+
+		// Associate with a layer if possible -- failing to do this will not affect
+		// the scan, just remove some traceability.
+		if !entry.EmptyLayer {
```
I absolutely agree! The only reason I didn't do this is to avoid repeating the assignment into the array and incrementing the layer index.

Keeping those separate means that there is less chance of introducing a bug (by omitting the increment or assignment) if there is additional logic added into this segment.

Regardless, however, I'd prefer to align with the preferences of those maintaining the code-base as it's really you that I'm trying to optimise for. 

With that in mind, given the above reasoning, what would you prefer to maintain? I'll happily align with your suggestion!

ahrav replied:
That’s a compelling reason. Let’s leave it as is. Thanks for the explanation.

jamestelfer replied:
And I really appreciate your thoughtful and clear commentary. 💪
